File: src/main/julia/aocd.jl
module Aocd

using Printf

function aocd_dir()
    try
        return ENV["AOCD_DIR"]
    catch KeyError
        try
            return joinpath(ENV["HOME"], ".config/aocd")
        catch KeyError
            throw(SystemError("aocd dir not found"))
        end
    end
end

function token()
    try
        return ENV["AOC_SESSION"]
    catch KeyError
        lines = readlines(joinpath(aocd_dir(), "token"))
        if isempty(lines)
            throw(SystemError("Missing session id"))
        else
            return lines[1]
        end
    end
end

function user_ids()
    lines = readlines(joinpath(aocd_dir(), "token2id.json"))
    if isempty(lines)
        throw(SystemError("Missing token2id.json"))
    else
        # Not ideal, but JSON is too slow...
        userIds = Dict()
        for line ∈ lines
            m = match(r""".*"(.*)": "(.*)",?""", line)
            if m != nothing
                userIds[m.captures[1]] = m.captures[2]
            end
        end
        return userIds
    end
end

function memo_dir()
    return joinpath(aocd_dir(), user_ids()[token()])
end

function day_format(year::Int, day::Int)
    return @sprintf("%d_%02d", year, day)
end

function get_input_data(year::Int, day::Int)
    return readlines(joinpath(memo_dir(),
                              day_format(year, day) * "_input.txt"))
end

function answer(year::Int, day::Int, partnum::Int)
    part = partnum == 1 ? "a" : "b"
    lines = readlines(joinpath(memo_dir(),
                               day_format(year, day) * part * "_answer.txt"))
    return lines[1]
end

function check(year::Int, day::Int, part1, part2)
    answer1 = answer(year, day, 1)
    fail1 = ""
    if !isempty(answer1) && !isempty(part1) && answer1 != string(part1)
        fail1 = string("Part 1: Expected: ", answer1, ", got: ", part1)
    end
    answer2 = answer(year, day, 2)
    fail2 = ""
    if !isempty(answer2) && !isempty(part2) && answer2 != string(part2)
        fail2 = string("Part 2: Expected: ", answer2, ", got: ", part2)
    end
    if !isempty(fail1) || !isempty(fail2)
        throw(AssertionError(join([fail1, fail2], '\n')))
    end
end

end  # module Aocd

File: src/main/cpp/aocd/aocd.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <regex>
#include "aocd.hpp"

using namespace std;

void aocd::sys::getFileContent(const string fileName, vector<string>& vec) {
    ifstream in(fileName);
    if (!in || !in.good()) {
        return;
    }
    string str;
    while (getline(in, str)) {
        vec.push_back(str);
    }
    in.close();
}

string aocd::sys::getAocdDir() {
    if (const char* env_d = getenv("AOCD_DIR")) {
        const string envs_d = env_d;
        if (!envs_d.empty()) {
            return envs_d;
        }
    } else {
        if (const char* env_h = getenv("HOME")) {
            const string envs_h = env_h;
            if (!envs_h.empty()) {
                return envs_h + "/.config/aocd";
            }
        }
    }
    throw "aocd dir not found";
}

string aocd::sys::getToken() {
    if (const char* env_t = getenv("AOC_SESSION")) {
        const string envs_t = env_t;
        if (!envs_t.empty()) {
            return envs_t;
        }
    } else {
        vector<string> vec;
        sys::getFileContent(sys::getAocdDir() + "/token", vec);
        if (!vec.empty()) {
            return vec[0];
        }
    }
    throw "Missing session ID";
}

map<string, string> aocd::sys::getUserIds() {
    const string fn = sys::getAocdDir() + "/token2id.json";
    vector<string> vec;
    sys::getFileContent(fn, vec);
    if (vec.empty()) {
        throw "Missing token2id.json";
    }
    regex re(".*\"(.*)\": \"(.*)\",?");
    map<string, string> m;
    for (const string& line : vec) {
        smatch sm;
        bool matched = regex_match(line, sm, re);
        if (matched) {
            m[sm[1]] = sm[2];
        }
    }
    return m;
}

string aocd::user::getMemoDir() {
    return sys::getAocdDir() + "/" + sys::getUserIds()[sys::getToken()];
}

vector<string> aocd::puzzle::getInputData(const uint year, const uint day) {
    char buf[255];
    const string fn = "%d_%02d_input.txt";
    sprintf(buf, &fn[0], year, day);
    vector<string> input;
    sys::getFileContent(user::getMemoDir() + "/" +  buf, input);
    if (input.empty()) {
        cerr << "!! INPUT DATA MISSING !" << endl;
    }
    return input;
}

string aocd::puzzle::answer(const uint year, const uint day, const uint partnum) {
    const string part = partnum == 1 ? "a" : "b";
    char buf[255];
    const string fn = "%d_%02d" + part + "_answer.txt";
    sprintf(buf, &fn[0], year, day);
    vector<string> lines;
    sys::getFileContent(user::getMemoDir() + "/" +  buf, lines);
    if (lines.empty()) {
        return "";
    } else {
        return lines[0];
    }
}

int aocd::puzzle::check(const uint year, const uint day, const string part1, const string part2) {
    const string& answer1 = puzzle::answer(year, day, 1);
    string fail1 = "";
    if (!answer1.empty() && !part1.empty() && answer1 != part1) {
        fail1 = "Part 1: Expected: " + answer1 + ", got: " + part1;
    }
    const string& answer2 = puzzle::answer(year, day, 2);
    string fail2 = "";
    if (!answer2.empty() && !part2.empty() && answer2 != part2) {
        fail2 = "Part 2: Expected: " + answer2 + ", got: " + part2;
    }
    if (!fail1.empty() || !fail2.empty()) {
        cerr << endl << "CHECK FAILED!!" << endl;
        cerr << fail1 << endl << fail2 << endl;
        return 1;
    }
    return 0;
}
